File: Algorithms/C++/416-Partition_Equal_Subset_Sum.cpp
// Runtime: 148 ms, faster than 37.45% of C++ online submissions for Partition Equal Subset Sum.
// Memory Usage: 8.4 MB, less than 82.35% of C++ online submissions for Partition Equal Subset Sum.
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum_half = accumulate(nums.begin(), nums.end(), 0);
        if (sum_half & 1)
            return false;
        else
            sum_half /= 2;
        vector<bool> dp(sum_half + 1);
        dp[0] = true;
        for (int num : nums)
            for (int i = sum_half; i >= num; --i)
                dp[i] = dp[i] || dp[i - num];
        return dp[sum_half];
    }
};

File: Algorithms/Python3.x/416-Partition_Equal_Subset_Sum.py
# dp
# Runtime: 680 ms, faster than 50.36% of Python3 online submissions for Partition Equal Subset Sum.
# Memory Usage: 13.2 MB, less than 43.64% of Python3 online submissions for Partition Equal Subset Sum.
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sum_half = sum(nums)
        if sum_half & 1: return False  # if sum is odd number return False
        else: sum_half = int(sum_half / 2)
        dp = [False] * (sum_half + 1)
        dp[0] = True
        for num in nums:
            for i in range(sum_half, num - 1, -1):
                dp[i] = dp[i] or dp[i - num]
        return dp[sum_half]


# dictionary
# Runtime: 1116 ms, faster than 34.80% of Python3 online submissions for Partition Equal Subset Sum.
# Memory Usage: 15.2 MB, less than 36.36% of Python3 online submissions for Partition Equal Subset Sum.
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        dict = {0}  # set
        for num in nums:
            dict.update({(tmp + num) for tmp in dict})
        return sum(nums) / 2.0 in dict
